File: apps/cli/src/commands/term.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import * as clack from "@clack/prompts";
import { getSession, setTerm } from "../auth/keychain.ts";
import { buildClient, AuthExpiredError, unwrapError } from "../api/client.ts";
import { listTerms, selectTerm } from "../api/terms.ts";

function requireSession() {
  const session = getSession();
  if (!session) {
    console.error(chalk.red("Not logged in. Run `uoplan login` first."));
    process.exit(1);
  }
  return session;
}

async function runInteractive(): Promise<void> {
  const session = requireSession();
  const spinner = ora("Fetching terms").start();

  let terms;
  try {
    const { client } = await buildClient(session);
    terms = await listTerms(client);
    spinner.stop();
  } catch (err) {
    spinner.fail();
    const e = unwrapError(err);
    if (e instanceof AuthExpiredError) {
      console.error(chalk.red(e.message));
      process.exit(1);
    }
    throw err;
  }

  if (terms.length === 0) {
    console.log(chalk.yellow("No terms available."));
    return;
  }

  const selected = await clack.select({
    message: "Select a term",
    options: terms.map((t) => ({
      value: t.index,
      label: t.name,
      hint: t.career,
    })),
  });

  if (clack.isCancel(selected)) {
    process.exit(0);
  }

  const spinner2 = ora(`Selecting ${terms[selected as number].name}`).start();

  try {
    const { client } = await buildClient(session);
    const strm = await selectTerm(client, selected as number);
    setTerm(strm, selected as number);
    spinner2.succeed(`Term set to ${chalk.bold(terms[selected as number].name)}.`);
  } catch (err) {
    spinner2.fail();
    const e = unwrapError(err);
    if (e instanceof AuthExpiredError) {
      console.error(chalk.red(e.message));
      process.exit(1);
    }
    throw err;
  }
}

async function runList(): Promise<void> {
  const session = requireSession();
  const spinner = ora("Fetching terms").start();

  try {
    const { client } = await buildClient(session);
    const terms = await listTerms(client);
    spinner.stop();

    for (const term of terms) {
      console.log(`${chalk.bold(term.name)}  ${chalk.dim(term.career)}`);
    }
  } catch (err) {
    spinner.fail();
    if (err instanceof AuthExpiredError) {
      console.error(chalk.red(err.message));
      process.exit(1);
    }
    throw err;
  }
}

export const termCommand = new Command("term")
  .description("Select a term interactively")
  .action(runInteractive)
  .addCommand(
    new Command("ls").description("List available terms (non-interactive)").action(runList),
  )
  .addCommand(
    new Command("list").description("List available terms (non-interactive)").action(runList),
  );

File: apps/cli/src/commands/term.rs
use anyhow::{anyhow, Result};
use cliclack::{intro, log, outro, outro_cancel, select, spinner};

use crate::api::terms::{list_terms, select_term};
use crate::api::PeopleSoftClient;
use crate::auth::{get_session, set_term};
use crate::error::NoCookiesError;

async fn fetch_terms() -> Result<(PeopleSoftClient, Vec<crate::api::peoplesoft::Term>)> {
    let session = get_session().await.ok_or_else(|| anyhow!(NoCookiesError))?;
    let client = PeopleSoftClient::new(session)?;
    let sp = spinner();
    sp.start("Fetching available terms…");
    let terms = list_terms(&client).await?;
    sp.stop("Terms loaded");
    Ok((client, terms))
}

pub async fn interactive() -> Result<()> {
    intro("uoplan term")?;
    let (client, terms) = fetch_terms().await?;
    if terms.is_empty() {
        outro_cancel("No terms available.")?;
        return Ok(());
    }

    let mut prompt = select("Select a term");
    for t in &terms {
        prompt = prompt.item(t.index, &t.name, &t.career);
    }
    let idx = match prompt.interact() {
        Ok(v) => v,
        Err(_) => {
            outro_cancel("Cancelled.")?;
            return Ok(());
        }
    };

    let chosen = terms.iter().find(|t| t.index == idx).unwrap();
    let sp = spinner();
    sp.start("Selecting term…");
    let strm = select_term(&client, idx).await?;
    sp.stop("Term selected");
    set_term(&strm, idx, None).await?;
    outro(&format!("Selected {}", chosen.name))?;
    Ok(())
}

pub async fn list() -> Result<()> {
    let (_client, terms) = fetch_terms().await?;
    if terms.is_empty() {
        log::info("No terms available.")?;
        return Ok(());
    }
    for t in &terms {
        log::info(&format!("{} — {}", t.name, t.career))?;
    }
    Ok(())
}
